src/note.ts

```diff
@@ -250,71 +494,65 @@ export class Note extends Tickable {
   }
 
   // Get and set the target stave.
-  getStave() {
+  getStave(): Stave | undefined {
     return this.stave;
   }
-  setStave(stave) {
+  setStave(stave: Stave): this {
     this.stave = stave;
     this.setYs([stave.getYForLine(0)]); // Update Y values if the stave is changed.
-    this.context = this.stave.context;
+    this.setContext(this.stave.getContext());
     return this;
   }
 
   // `Note` is not really a modifier, but is used in
   // a `ModifierContext`.
-  getCategory() {
+  getCategory(): string {
     return Note.CATEGORY;
   }
 
-  // Set the rendering context for the note.
-  setContext(context) {
-    this.context = context;
-    return this;
-  }
-
   // Get and set spacing to the left and right of the notes.
-  getLeftDisplacedHeadPx() {
+  getLeftDisplacedHeadPx(): number {
     return this.leftDisplacedHeadPx;
   }
-  getRightDisplacedHeadPx() {
+  getRightDisplacedHeadPx(): number {
     return this.rightDisplacedHeadPx;
   }
-  setLeftDisplacedHeadPx(x) {
+  setLeftDisplacedHeadPx(x: number): this {
     this.leftDisplacedHeadPx = x;
     return this;
   }
-  setRightDisplacedHeadPx(x) {
+  setRightDisplacedHeadPx(x: number): this {
     this.rightDisplacedHeadPx = x;
     return this;
   }
 
   // Returns true if this note has no duration (e.g., bar notes, spacers, etc.)
-  shouldIgnoreTicks() {
+  shouldIgnoreTicks(): boolean {
     return this.ignore_ticks;
   }
 
   // Get the stave line number for the note.
-  getLineNumber() {
+  getLineNumber(): number {
     return 0;
   }
 
   // Get the stave line number for rest.
-  getLineForRest() {
+  getLineForRest(): number {
     return 0;
   }
 
   // Get the glyph associated with this note.
-  getGlyph() {
+  getGlyph(): Glyph | GlyphProps {
     return this.glyph;
   }
 
-  getGlyphWidth() {
+  getGlyphWidth(): number {
     // TODO: FIXME (multiple potential values for this.glyph)
     if (this.glyph) {
       if (this.glyph.getMetrics) {
         return this.glyph.getMetrics().width;
-      } else if (this.glyph.getWidth) {
-        return this.glyph.getWidth(this.render_options.glyph_font_scale);
+        //} else if (this.glyph.getWidth) {
+        //  return this.glyph.getWidth(this.render_options.glyph_font_scale);
       }
     }
```

Comment:
@0xfe I guess that I should just use any, should I?

 But this is a topic worth investigating. Do you know anyone that could take over this issue?

Reply:
We need that `else` for sure. What was the compiler error? Doesn't `getWidth` return `number`?

Reply:
Either way, okay to use `any` here with `eslint-disable` and maybe a TODO because we know we need to fix it.

Reply:
```
ERROR in /home/rodrigo/git/vexflow/src/note.ts
./src/note.ts 554:28-36
[tsl] ERROR in /home/rodrigo/git/vexflow/src/note.ts(554,29)
      TS2339: Property 'getWidth' does not exist on type 'Glyph | GlyphProps'.
  Property 'getWidth' does not exist on type 'Glyph'.

ERROR in /home/rodrigo/git/vexflow/src/note.ts
./src/note.ts 555:26-34
[tsl] ERROR in /home/rodrigo/git/vexflow/src/note.ts(555,27)
      TS2339: Property 'getWidth' does not exist on type 'Glyph | GlyphProps'.
  Property 'getWidth' does not exist on type 'Glyph'.
```

Reply:
Okay, yeah, you need to do a type check, something like (off the top of my head, not sure if it compiles, but this is the general idea):

```typescript
else if (typeof this.glyph == 'Glyph') {
  return (this.glyph as Glyph).getWidth(this.render_options.glyph_font_scale);
}
```

Reply:
The error is complaing about the fact that one of the members in the double type, `Glyth` does not have the function `getWidth` I have defined now the function member in `Glyth`. 

But I have also realised that both types implement `getMetric` so I believe that this else will never be executed.

Now it is compiling and running the tests correctly.

Reply:
If you look at `tables.js` there are some data objects that implement a custom `getWidth`, so I think this may not be sufficient. I think we should create an interface:

```typescript
interface WidthGetter {
  getWidth(): number
}
```

Type `glyph` as:

```typescript
glyph: Glyph | GlyphProps | WidthGetter
```

And make this line:

```typescript
else if ('getWidth' in glyph as any) {
  return (this.glyph as WidthGetter).getWidth(this.render_options.glyph_font_scale);
}
```

Reply:
The compiler will then complain about WidthGetter not implementing `getMetrics`. 

I do not know, I do not think that we should twist TS to comply with the code, we should untwist the code with the help of TS. Wouldn't you agree? I would rather leave any on the things we do not figure out how to simplify (this also applies to formatter)

Reply:
Are you sure? It shouldn't complain about that because `glyph` is a union type that includes `Glyph | GlyphProps | WidthGetter`.

I'm fine with using `any` here too.

Reply:
The error was:
```
[tsl] ERROR in /home/rodrigo/git/vexflow/src/note.ts(555,27)
      TS2339: Property 'getWidth' does not exist on type 'Glyph | GlyphProps'.
  Property 'getWidth' does not exist on type 'Glyph'.
```
the compound type does not support `getWidth` and then the detail `Glyph` does not have it.

I have included `getWidth` in `Glyph` and now all the tests are runing.

I believe leaving it like it is now will force a check on the next ts migrated files that it is either `Glyph | GlyphProps` and we will see if something else is popping up. For the js file it does not matter because there is no type checking on run time.

Reply:
Thanks, right, I understand why the error is happening. We can either use the interface or use `any` types, both which are permissive. But I'm not very comfortable with this approach because I know for certain that it doesn't cover all cases. Also, we shouldn't need to add `getWidth` to `Glyph` to make this work -- it's difficult to reason about that implementation change.

Let's leave `any` and deal with this separately, it's a reasonable place for it.

Reply:
Left as any